### ISSUE TITLE
Unsafe Sharp Turns?

### ISSUE THREAD

**Issue by DicePenguin** [organicmaps/organicmaps#8637]
⚠ Have you searched for similar, already existing issues?
Yes

**Describe the issue**
The navigation seems to have a habit of suggesting very sharp turns. Is there a way to filter out most of these without having to add turn restrictions to every turn on openstreetmap.org? Currently, every time I drive somewhere new, one of these seems to come up and I have to add turn restrictions for it in openstreetmap.org.

**Steps to reproduce**
1. Navigate to a place.
2. Wait for sharp turn to be suggested off or onto an exit or some other highway.

**Expected behaviour**
I would expect the navigation to avoid very sharp turns.

Please see screenshot below:
![Screenshot](https://github.com/organicmaps/organicmaps/assets/84756738/f447da63-723e-4926-b60b-3e1dac5d1db7)

**System information:**
 - Operating system: iOS 17.4.1
 - Organic Maps:        2024.06.19-4
 - Device Model:         iPhone XR

**Comment by biodranik** [comment]
Is the issue about _wrong_, _prohibited_ turns? If yes, then it's map data issue. Organic Maps can filter them out too, but then some valid sharp turns can also be filtered out.

**Comment by RedAuburn** [comment]
data issue: https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=40.4016%2C-74.3064%3B40.4078%2C-74.3055#map=17/40.40468/-74.30482

although as said before, Valhalla & mapbox can handle this properly without explicit tagging: https://github.com/organicmaps/organicmaps/issues/8097

related
https://github.com/organicmaps/organicmaps/issues/4290

**Comment by DicePenguin** [comment]
RedAuburn seems to have summed up the problem by mentioning those other two issues.

**Comment by biodranik** [comment]
An ideal solution is to suggest users to fix map data issues semi-automatically, without learning how to edit relations.

**Comment by RedAuburn** [comment]
> An ideal solution is to suggest users to fix map data issues semi-automatically, without learning how to edit relations.

i agree in principle, but OM already prohibits lots of edit types due to being too risky, routing edits are even more risky than what is currently prohibited

**Comment by biodranik** [comment]
Is it possible to make a check/logic like this? Which risks do you see?

1. Check if there's any existing relation/restriction.
2. If there's none, ask the user if it's allowed to turn or not.
3. If the user answers No, then the code _properly_ edits OSM data by adding a restriction.

**Comment by DicePenguin** [comment]
I think the benefits of reducing very sharp turns (especially left turns in the US) outweighs the risks of removing a few legitimate sharp turns. Can we place a filter on how acute the turn angle is (for example 15 degrees and smaller)?

**Comment by AtmosphericIgnition** [comment]
Hi, I actually had [edited](https://www.openstreetmap.org/changeset/161878114) OSM to get around an issue where I was being routed as shown by the red line, but all that resulted in the OM making a route that just doesn't make any sense, since it would have you take a left through a signaled intersection, take another left across a busy road only to turn around in a parking lot. The second screenshot shows much saner routes, which are actually faster, since they follow major road, instead of taking a double left turn, and snaking through a busy downtown area.

![Image](https://github.com/user-attachments/assets/9163f0d2-3856-4905-b587-5321dc08647f)

Significantly safer and faster routes:
![Image](https://github.com/user-attachments/assets/caa1d81d-6fd8-4a33-bea4-2d19cdc452d5)

**Comment by dinabarysevich** [comment]
<img width="1080" height="2400" alt="Image" src="https://github.com/user-attachments/assets/90a00614-d65b-4f7e-b9ab-6ff496beab6b" />

The user complained on a sharp turn. 

Ways to reproduce: the route was started after the exit to the highway. Therefore, OM calculated the fastest route based on speed tags and road types.
Since there is no left turn restriction in the OSM data, OM considers this to be the fastest route (not prohibited = permitted).

Can be fixed by adding a turn restriction in OSM.

https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=48.54075%2C35.05651%3B48.53681%2C35.05527#map=16/48.53902/35.05718
